To-do 리스트 Redux를 사용하여 리팩터링 하기

yxxnhx opened the pull request:
# 목표
To-do 리스트 Redux를 사용하여 리팩터링 하기

## 진행사항
3일만에 첫 PR을 남깁니다..😢
리덕스를 처음 사용해보는 것이라 강의를 계속 여러번 들으면서 정리하고 모르는 개념은 찾아보면서 하다보니 조금 시간이 오래 걸렸네요
더 늦기전에라도 올려야 할 것 같아서 리덕스를 활용하여 reduce 테스트 파일까지 완료한 상태로 올립니다. 
사실은 야살님의 강의를 듣고 필기한 것을 계속 참고하며 진행하고 있습니다🥲
혹시 리덕스에 대해서 더 참고할 수 있는 사이트가 있다면 추천해주시면 정말 감사하겠습니다!
그리고 지금 계속 테스트 커버리지가 100%를 못채워서 E2E가 깨지네요ㅠㅠ 
얼른 더 부지런히 진행해보도록 하겠습니다💪

yxxnhx committed:
App 컴포넌트 내에서 관심사 분리를 하였습니다.

상태를 관리하는 함수 updateTaskTitle, addTask, deleteTask를 생성하여 App 컴포넌트 외부로 분리시켰습니다.
₩

yxxnhx committed:
리덕스를 활용하여 관심사 분리를 하였습니다

action creator는 actions에, 전체적인 상태관리는 store.js를 생성하여 관심사에 따른 분리시켰습니다

yxxnhx committed:
store에서 reducer을 분리한 후 test를 진행하였습니다.

각각의 action에 따라 분리하여 test를 진행하였습니다.

yxxnhx committed:
reducer의 tasks 기본값을 삭제하였습니다

test용으로 fixtures에 분리하여 사용하므로 기본 세팅값이 필요하지 않아 삭제하였습니다

yxxnhx committed:
reducer test 파일 중 deleteTask 테스트명을 수정하였습니다.

id가 있을 경우, 없을 경우를 나누어서 진행할 수 있도록 수정했습니다.

yxxnhx committed:
reducer과 reducer test에 아무런 값이 없을 경우의 케이스를 추가했습니다

reducer에서 return state인 상태로 아무런 값이 없을 경우 아무런 일도 일어나지 않음을 추가하면 action type을 찾지 못하는 에러가 계속해서 발생했습니다.
따라서 reducer에서 actionCreators를 이용하여 각각의 케이스를 오브젝트로 받아 아무런 값이 없을 경우의 케이스를 추가하였습니다.

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,82 @@
+
+        expect(state.tasks).toHaveLength(1);
+      });
+    });
```
혹시, 이 테스트는 tasks에 존재하지 않는 id일 경우, 아무런 동작을 하지 않는 걸 기대하고 한 테스트인가요?
만약, 그렇다면, 숫자를 정말 존재할 수 없는 `음수`로 값을 넣어줄 것 같아요! ㅎㅎ진짜 존재하지 않는 숫자라는 것을 드러내기 위해서요!

yxxnhx replied:
아하 막연히 그냥 다른 숫자면 된다고 생각했는데 정말 존재할 수 없는 숫자를 이용하는 편이 훨씬 더 정확하겠네요! 변경하겠습니다~!

daadaadaah commented on src/reducer.test.js:
```diff
@@ -0,0 +1,82 @@
+import { updateTaskTitle, addTask, deleteTask } from './actions';
+
+describe('reducer', () => {
+  describe('updateTaskTitle를 실행할 시', () => {
```
describe는 `테스트 대상`에 대해 것이므로, 그냥 `updateTaskTitle`로 기재해주셔도 됩니다!ㅎㅎ(다른 테스트 코드도 같이 바꿔주세요!)

yxxnhx replied:
넵 전체적으로 수정하겠습니다!

daadaadaah commented on src/reducer.js:
```diff
@@ -0,0 +1,40 @@
+      ...state,
+      tasks: tasks.filter((task) => task.id !== action.payload.id),
+    };
+  }
```
우리는 이전 과제들을 하면서 if문이 많을 경우, 어떤 걸 활용해서 코드를 간단히 했습니다!ㅎㅎ한번 적용시켜봅시당!ㅎㅎ

daadaadaah replied:
이 코멘트는 무시해주세요! 이미 잘 적용시키셨네용!👏👏👏👏👏👏

daadaadaah commented on __mocks__/react-redux.js:
```diff
@@ -0,0 +1,3 @@
+export const useDispatch = jest.fn();
+
+export const useSelector = jest.fn((selector) => selector({}));
```
아샬님은 useSelector를 어떻게 mocking 하셨는지 강의를 통해 한번 체크해봅시당!ㅎㅎ

yxxnhx replied:
앗 더는 기본값이 필요없네요 수정하겠습니다!

daadaadaah commented on src/App.test.jsx:
```diff
@@ -1,14 +1,22 @@
+
 import App from './App';
 
+import tasks from '../fixtures/tasks';
```
오! 앞에서 배운대로 바로 `fixtures`를 활용하셨군요! 잘하셨어요👏

daadaadaah reviewed (changes requested):
redux 처음 접하시면 어려우실텐대 3일만에 PR 날리신 점 너무 잘하셨어요!👍👍👍  
남겨드림 코멘트 참고해서 과제를 마저 완성시켜 봅시다! 그리고, 의도적으로 TDD 싸이클에 맞춰 코딩해보려고 노력해보세요! 지금 습관을 잘 들여놔야 나중에 자연스럽게 TDD 하실 수 있어요! 작업 단위도 작게 해보시구요!ㅎㅎ그럼 다음 PR 기다리고 있을게요!🚀

> 혹시 리덕스에 대해서 더 참고할 수 있는 사이트가 있다면 추천해주시면 정말 감사하겠습니다!
- 제가 공부했을 떄에는 생활코딩 참고했었어요! 리액트 11기분들도 생활코딩 강의 이해하기 쉽다고 하시더라구요!
- 저는 아래 링크 보면서 직접 손으로 그려보기도 하면서 redux를 맨처음에 이해한 것 같아요! 도움이 되었으면 좋겠습니다😊
참고 : https://www.youtube.com/playlist?list=PLuHgQVnccGMB-iGMgONoRPArZfjRuRNVc

yxxnhx committed:
reducer test의 id가 존재하지 않을 경우 테스트 조건 수정하였습니다.

정말 존재할 수 없는 숫자인 음수로 변경하여 테스트 코드를 수정하였습니다.

yxxnhx committed:
reducer test decribe 테스트 이름을 수정하였습니다.

테스트 대상으로만 모두 통일하여 수정하였습니다.

yxxnhx committed:
ListContainer, ListContainer test 파일을 생성하였습니다.

목록에 대한 관심사를 분리하여 목록에 관련된 tasks, deleteTask를 이용하여 생성하였습니다.

yxxnhx committed:
InputContainer, InputContainer.test 파일을 생성하였습니다.

입력과 관련된 taskTitle과 updateTaskTitle, addTask를 분리하여 InputContainer를 생성하였습니다.

yxxnhx committed:
관심사 분리를 통한 App, App.test 정리 및 Page, Page.test 삭제

App에서 그냥 Page로 넘기고 Page에서 InputContainer, ListContainer를 넘기는 무의미한 상황이 반복되어 App에서 바로 넘기는 로직 변경으로 Page, Page.test를 삭제하였습니다

yxxnhx committed:
ListContainer.test에서 할일 랜더링, 완료 버튼 호출 테스트 케이스를 추가하였습니다.

입력한 할일을 랜더링, 선택한 id에 맞는 완료 deleteTask가 호출되는지 확인할 수 있는 테스트 코드를 추가하였습니다.

yxxnhx committed:
InputContainer.test 테스트 케이스를 추가하였습니다.

InputContainer의 랜더링, 입력한 할일의 랜더링, 추가 버튼을 누르면 addTask 호출을 확인할 수 있는 케이스를 추가하였습니다.

yxxnhx committed:
mocks의 useSelector를 수정하였습니다.

기본값 없이도 충분히 사용할 수 있으므로 useSelector의 기본값을 삭제하였습니다.

yxxnhx committed:
반복되어 출력된 To-do 타이틀을 수정하였습니다.

파일을 분리하면서 App 컴포넌트와 InputContainer 컴포넌트에 모두 입력되어있어 중복되어 출력되는 부분을 삭제하였습니다.

yxxnhx commented:
안녕하세요 트레이너님!
트레이너님 조언을 참고해서 test의 id가 존재하지 않을 경우의 테스트 조건을 음수로 수정하여 정말 존재하지 않을 수 있도록 변경하였습니다.
그리고 전체적으로 필요없는 말이 들어간 descirbe도 수정하였습니다.
관심사 분리를 위해 ListContainer와 InputContainer를 생성하여 각각의 테스트 케이스를 추가하여 테스트 커버리지 100을 달성하였습니다.
아직 많은 것이 부족해 이것저것 찾아보면서 하고 있는 중입니다💪
트레이너님께서 추천해주신 생활코딩 리덕스 강의를 참고해서 더 해보도록 하겠습니다!
강의 추천해주셔서 감사합니다❤️‍🔥

daadaadaah reviewed (approved):
잘하셨어요!👏 이제 commit를 명령형으로 써보시는건 어떠신가요? `~ 수정했습니다` 보다 보통 `~ 수정` 간단히 적어줍니다!ㅎㅎ과제2도 화이팅해봐요!💪
